_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,174 @@
+
+* A JSON file of this magnitude takes a fair amount of time to download.
+
+## The Solution:
+
+* _Remove extraneous Data_ from the array and ensure this operation only happens once
+
+* _Create a seperate thread to update indexedDB_, while the user progressed with other aspects of the application.
+
+* _Improve the download speed of the data for slow networks_ by chunking requests in parallel
+_(optional and beyond the scope of this post)_
+
+1. *Remove extraneous Data*
+
+```typescript
+async function transformPokemon(): Promise<void> {
+    for (let i = 0; i < this.Pokemon.length; i++) {
+        const pokemon = {
+            value: this.pokemon[i].pokemonId,
+            label: `${this.pokemon[i].pokemonId} - ${this.pokemon[i].shortDescription}`
+        };
+        this.pokemonOptions.push(pokemon);
+    }
+    await localforage.setItem(this.pokemonGroup, this.pokemonOptions);
+}
+```
+
+Instead of storing the complete dataset in memory and using it in several different places. We pulled out only the data that was absolutely necessary for displaying in the UI. We also made sure that this array mapping operation only happened once every time the UPDATE_POKEMON action was dispatched to our redux store, instead of in a component render method. You'll notice, we also store a copy of this transformed array so that we don't have to retrieve it over the network the next time. Which means we can run the following when mounting our component.
+
+```typescript
+async function getPokemon(): Promise<AnyAction> {
+    const pokemonDAO = new PokemonDAO(this.location);
+    const localPokemonCount = await pokemonDAO.getCount();
+
+    if (localPokemonCount > 0) {
+        // The data has already been retrieved
+        this.pokemonOptions = await localforage.getItem(this.pokemonGroup) as SelectOption[];
+    } else {
+        // fetch the data and update IndexedDB
+        const options = { method: 'GET' };
+        this.pokemon = await new NetworkCall(this.endpoint, options).makeRequest() as Pokemon[];
+        this.spawnWorker();
+        await this.transformPokemon();
+    }
+    return this.buildUpdateAction();
+}
+```
+2. *Create a seperate thread in the background*
+
+Before we completed the above step, we decided to do something a little less obvious that would eventually aid us in our quest to build an offline web app. We decided to create a local Database of all the Pokemon retrieved from the network, in IndexedDB. In case you don’t know, IndexedDB is a native browser API that is supported by most modern browsers.
+
+When it came to inserting the data into a table in IndexedDB, we found that the native javascript/browser IndexedDB API was cumbersome and unintuitive. We found a wrapper called `Dexie` on NPM (Pete is particularly fond of the name of this module). Dexie also allowed us to perform bulk  put operations on our IndexedDB tables by opting out of listening to onSuccess callbacks for each put operation.
```

Comment:
- Suggest hyperlinking to Dexie.
- What's the value of throwing pete under the bus? haha
- two spaces between `bulk` and `put`
- Suggest adding why this was advantageous at the end regarding bulk operations

Reply:
Who is this Pete you speak of?